sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -363,6 +364,70 @@ trait Row extends Serializable {
     false
   }
 
+  protected def canEqual(other: Any) =
+    other.isInstanceOf[Row] && !other.isInstanceOf[InternalRow]
```

Comment:
Maybe we can create a new method for the comparison, and within `equals` method, we can check the null/type  before invoking the comparison method. So we can remove the `canEqual` method if it work in this way.